### PR TITLE
Feature/add multiselect filter

### PR DESCRIPTION
**Pull request by cassmbautista**
Added feature #405

This new feature is to add a multiple selection option to vue-good table. Seen in the image below:

2019-08-06 at 12 27 PM

The changes made in this PR are:
* Added library for multiselect options called: [vue-select](https://github.com/sagalbot/vue-select) 
* Using new component added an option to the filter options to set the multiselect dropdown. 
  * This option does a fuzy search on any of the selected options. 
  * This feature will work with the specified types as well 
  * This also works when reseting the table
* Updated docs to include documentation for new filter options
* Updated dev app to use new feature

### CODE REVIEW

**Comment by xaksis** [comment]
This is sweeet! @cassmbautista my only concern is adding another library for something that is not core to the table and might not be used by everyone.
I'm trying to trim down the existing ones to reduce the size (lodash etc). Is it possible to add this as a plugin for vue-good-table? that way you can maintain it and people can optionally add it if they want to use it?

**Comment by cassmbautista** [comment]
@xaksis Yea I was concerned about that too since I noticed you have very little dependencies. However, I have no idea how to add it as a plugin for vue-good-table. Any ideas? Do you have any examples of plugins to components?

**Comment by xaksis** [comment]
Good point. I was actually thinking of making a column-manager plugin for VGT myself to figure out how that'd work. Mind if I hold off on this for a bit? Then we can use that as a template if it works.

**Comment by cassmbautista** [comment]
Yea, you can hold off on this for a while. I am also interested in the functionality of the column-manager plugin.

**Comment by TheJaredWilcurt** [comment]
@xaksis 
I was wanting a column manager option as well. My design for it would have been a button or icon that opens a pop-out menu with all columns listed in a scrollable area. Each column name would have a grabber icon on the left to drag it up/down to reorder them and a checkmark to the right of the column name to show/hide that column.

My original mockup:

mockup

The default sort allows the user to pick which column to sort by by default.

**Comment by xaksis** [comment]
haha @TheJaredWilcurt that's pretty much exactly what I have implemented in the project that I was planning to refactor. Glad to see it validated!

**Comment by bhargavjoshi** [comment]
Guys (@cassmbautista, @xaksis) while you are discussing the column plugin, it can be a good idea to give an autocomplete function in the filter (let say you want to filter by a particular user, and there are 1000s of the user and autocomplete server fetching is a good option) for those using remote mode

**Comment by TheJaredWilcurt** [comment]
@bhargavjoshi 
This PR uses `vue-select` which has that capability. I'm not sure if this PR supports that, but it could be an additional follow up PR to add that functionality if it doesn't.

* https://vue-select.org/guide/ajax.html

**Comment by bhargavjoshi** [comment]
@cassmbautista, @xaksis 

I was thinking about giving users the ability to use their own filter component. 


Right now we have these options. 
```
filterOptions: {
  enabled: true,
  filterValue: '',
  placeholder: "Show All",
  filterDropdownItems: ['']
}
```

We can add a property for a component like this 

```
filterOptions: {
  enabled: true,
  filterValue: '',
  component: '<MyCustomComponent :attr1="attr1" @change="functionToSetFilterValue" />'
}
```

We have to think about how we can pass data to the component and also get data from the component. 

My use case Is I want to give two text inputs with labels to put **min** amount and **max** amount. (basically a range filter). 

Let's discuss more on this.

**Comment by paulhuisman** [comment]
Nice work! Would love to see this get integrated!

**Comment by cassmbautista** [comment]
@xaksis Have you been able to make an example plugin for vue-good-table, so that I can convert this PR to a plugin?

**Comment by unixconky** [comment]
Would love to see this get integrated too! :)

**Comment by xaksis** [comment]
@cassmbautista really sorry but I have been very busy with another project. When I get a breather I'll get on it! I will keep you updated.

**Comment by unixconky** [comment]
It's looking good. Can't wait to see it integrated or as a plugin :)

**Comment by talbotseb** [comment]
Hi @xaksis ! 
Thanks a lot for your good work even if it lets us a bit dependant ...
Any idea about when you'll be able to come back on this project ?
It'll be nice to get some extra-features. If you're busy, maybe there is somebody here who could help you ?

Thanks anyhow !

File: src/components/Table.vue [hunk 1]
```diff
@@ -1228,6 +1228,25 @@ export default {
                     this.columnFilters[col.field]
                   );
                 }
+
+                // If the column has an array of filter values match any
+                if (col.filterOptions && col.filterOptions.filterMultiselectDropdownItems) {
+                  if(this.columnFilters[col.field].length === 0) {
```
Comment (TheJaredWilcurt) [inline src/components/Table.vue]:
```suggestion
                  if (this.columnFilters[col.field].length === 0) {
```

File: src/components/Table.vue [hunk 2]
```diff
@@ -1228,6 +1228,25 @@ export default {
...
+                    return true;
+                  }
+                  // Otherwise Use default filters
+                  const { typeDef } = col;
+                  for(let filter of this.columnFilters[col.field]) {
```
Comment (TheJaredWilcurt) [inline src/components/Table.vue]:
```suggestion
                  for (let filter of this.columnFilters[col.field]) {
```

File: src/components/Table.vue [hunk 3]
```diff
@@ -1228,6 +1228,25 @@ export default {
...
+                    if(typeDef.filterPredicate(
```
Comment (TheJaredWilcurt) [inline src/components/Table.vue]:
```suggestion
                    if (typeDef.filterPredicate(
```

File: src/components/VgtFilterRow.vue [hunk 4]
```diff
@@ -41,6 +41,17 @@
           :key="i"
           :value="option.value">{{ option.text }}</option>
       </select>
+
+      <!---->
```
Comment (TheJaredWilcurt) [inline src/components/VgtFilterRow.vue]:
```suggestion
```

File: src/components/VgtFilterRow.vue [hunk 5]
```diff
@@ -119,6 +141,11 @@ export default {
         && typeof column.filterOptions.filterDropdownItems[0] !== 'object';
     },
 
+    isMultiselectDropdown(column) {
+      return this.isFilterable(column)
```
Comment (TheJaredWilcurt) [inline src/components/VgtFilterRow.vue]:
```suggestion
      return this.isFilterable(column) &&
```

File: src/components/VgtFilterRow.vue [hunk 6]
```diff
@@ -119,6 +141,11 @@ export default {
...
+        && column.filterOptions.filterMultiselectDropdownItems;
```
Comment (TheJaredWilcurt) [inline src/components/VgtFilterRow.vue]:
```suggestion
        column.filterOptions.filterMultiselectDropdownItems;
```

File: src/components/VgtFilterRow.vue [hunk 7]
```diff
@@ -93,6 +103,17 @@ export default {
...
       this.columnFilters = {};
+
+      // Clear the selection in the multiselect
+      let multiselectRefs = this.columns
+        .map((column, index) => {
+          column.ref = 'vgt-multiselect' + column.label + index;
+          return column;
+        })
+        .filter(this.isMultiselectDropdown)
+        .map(column => column.ref);
```
Comment (TheJaredWilcurt) [inline src/components/VgtFilterRow.vue]: This can probably be a `filter > map` instead of `map > filter > map`

File: vp-docs/guide/configuration/column-filter-options.md [hunk 8]
```diff
@@ -61,6 +62,17 @@ filterDropdownItems: [
...
 
+## filterMultiselectDropdownItems
+
+type `Array of strings`
+
+allows creating a dropdown for filtering multiple items as opposed to an input
+
+```javascript
+//array
+filterMultiselectDropdownItems: ['Blue', 'Red', 'Yellow']
```
Comment (TheJaredWilcurt) [inline vp-docs/guide/configuration/column-filter-options.md]: add examples show using objects with label/id

**Comment by TheJaredWilcurt** [comment]
size-increase

Merging this is means a 15% increase in the payload size (since it will now include the vue-select dependency). If this can be set up as an optional dependency or to be tree-shakable that would be ideal. However I'm also fine with that being a follow up issue.

File: src/components/Table.vue [hunk 9]
```diff
@@ -1175,6 +1175,33 @@ export default {
...
+      if (column.filterOptions && column.filterOptions.filterMultiselectDropdownItems) {
+        if (columnFilters.length === 0) {
+          return true;
+        }
+
+        // Otherwise Use default filters
+        const typeDef = column.typeDef;
+        for (let filter of columnFilters) {
+          let filterLabel = filter;
+          if(typeof filter === 'object'){
```
Comment (TheJaredWilcurt) [inline src/components/Table.vue]:
```suggestion
          if (typeof filter === 'object') {
```

File: src/components/Table.vue [hunk 10]
```diff
@@ -1228,6 +1255,12 @@ export default {
                     this.columnFilters[col.field]
                   );
                 }
+
+                const filterMultiselect = this.filterMultiselectItems(col, row);
+                if(filterMultiselect !== undefined) {
```
Comment (TheJaredWilcurt) [inline src/components/Table.vue]:
```suggestion
                if (filterMultiselect !== undefined) {
```

File: src/components/VgtFilterRow.vue [hunk 11]
```diff
@@ -119,6 +135,11 @@ export default {
         && typeof column.filterOptions.filterDropdownItems[0] !== 'object';
     },
 
+    isMultiselectDropdown(column) {
+      return this.isFilterable(column) &&
+        column.filterOptions.filterMultiselectDropdownItems;
```
Comment (TheJaredWilcurt) [inline src/components/VgtFilterRow.vue]:
```suggestion
        column.filterOptions &&
        column.filterOptions.filterMultiselectDropdownItems
      );
```

File: src/components/VgtFilterRow.vue [hunk 12]
```diff
@@ -119,6 +135,11 @@ export default {
         && typeof column.filterOptions.filterDropdownItems[0] !== 'object';
     },
 
+    isMultiselectDropdown(column) {
+      return this.isFilterable(column) &&
```
Comment (TheJaredWilcurt) [inline src/components/VgtFilterRow.vue]:
```suggestion
      return (
        this.isFilterable(column) &&
```
